static/components.interfaces.ts

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) 2021, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+export interface ComponentConfig<S> {
+    type: string;
+    componentName: string;
+    componentState: S;
+}
+
+type StateWithLanguage = {lang: string};
+type StateWithEditor = {source: number};
+type StateWithTree = {tree: number};
+type StateWithId = {id: number};
+type EmptyState = Record<never, never>;
+
+export type EmptyCompilerState = StateWithLanguage & StateWithEditor;
+export type PopulatedCompilerState = StateWithEditor & {
+    filters: Record<string, boolean>; // TODO: compilation filters
```

Comment:
Thanks for leaving a TODO here; best to get something out that can be improved upon than obsessing over the minutiae. Thanks!

Reply:
filters are defined here https://github.com/compiler-explorer/compiler-explorer/blob/main/types/features/filters.interfaces.ts#L25 - though not sure if they are equal to the ones in our UI